Feature / CONTENTBOX-1296 - History Relocations

jclausen opened the pull request:
Resolves CONTENTBOX-1296.  Video Attached to demonstrate UI functionality.

https://user-images.githubusercontent.com/5255645/195915356-7b15acbd-2a35-4a8a-9edb-5610110aa40d.mov

lmajano commented on modules/contentbox/migrations/2022_10_13_133204_CONTENTBOX-1296-Unique-Site-Slug-Relocation.cfc:
```diff
@@ -0,0 +1,26 @@
+component {
+
+	function up( schema, qb ){
+		if ( schema.hasTable( "cb_relocations" ) ) {
```
Can you verify this actually works? In my tests, none of the `hasColumn and hasTable()` methods don't he schema builder actually worked.  That's why the migration utilities.cfm has some of these methods.

lmajano commented on modules/contentbox/models/content/ContentService.cfc:
```diff
@@ -1195,6 +1196,10 @@ component extends="cborm.models.VirtualEntityService" singleton {
 			.list( sortOrder = arguments.sortOrder );
 	}
 
+	function addRelocation( required BaseContent contentItem, required string originalSlug ){
```
Please add CFC docs

lmajano replied:
Also make it return `this` so we keep the API fluent

lmajano commented on modules/contentbox/models/content/RelocationService.cfc:
```diff
@@ -0,0 +1,127 @@
+		return results;
+	}
+
+	Relocation function createContentRelocation( required BaseContent contentItem, required string originalSlug ){
```
Add cfc docs please

lmajano commented on modules/contentbox/models/content/RelocationService.cfc:
```diff
@@ -0,0 +1,127 @@
+		return relocation;
+	}
+
+	Relocation function createTargetRelocation( required string slug, required string target ){
```
CFC Docs please

lmajano reviewed (changes requested):
still in progress, but adding some comments

lmajano commented on modules/contentbox/modules/contentbox-ui/handlers/blog.cfc:
```diff
@@ -251,6 +252,19 @@ component extends="content" {
 		} else {
+			var relocation = variables.relocationService
+				.newCriteria()
+				.isEq( "site", variables.CBHelper.site() )
```
Just use `prc.oCurrentSite` it's already been discovered

lmajano commented on modules/contentbox/models/system/CBHelper.cfc:
```diff
@@ -2563,4 +2563,17 @@ component accessors="true" singleton threadSafe {
+	 *
+	 * @obj any  the object to be serialized
+	 */
+	string function toJSON( any obj ){
```
ColdBOx already has a `forAttribute()` function to deal with CF -> JS apps. I think this is not necesasry

lmajano replied:
I would prefer we update the `forAttribute()` function in the framework itself.

jclausen replied:
Actually `forAttribute` does not work for this use case, as it also runs the JSON through `encodeForHTMLAttribute`, which is not what we want in this case - as it is being presented in a `script` block.  If we do modify the core, I would suggest adding this method, and then have `forAttribute` use the new method for its JSON

jclausen replied:
Created: https://ortussolutions.atlassian.net/browse/COLDBOX-1159

jclausen replied:
Updated to use new `toJSON` method in supertype/Util and bumped Coldbox to `be` until the next version is released.

jclausen commented on modules/contentbox/modules/contentbox-ui/handlers/blog.cfc:
```diff
@@ -251,6 +252,19 @@ component extends="content" {
 				.setLayout( name = "#prc.cbTheme#/layouts/blog", module = prc.cbThemeRecord.module )
 				.setView( view = "#prc.cbTheme#/views/entry", module = prc.cbThemeRecord.module );
 		} else {
+			var relocation = variables.relocationService
```
Good catch.  On it.

jclausen commented on modules/contentbox/modules/contentbox-admin/handlers/baseContentHandler.cfc:
```diff
@@ -275,6 +275,8 @@ component extends="baseHandler" {
 				return !reFindNoCase( "#prc.oContent.getSlug()#\/?", arguments.item[ "slug" ] );
 			} );
+		// Provide JWT Tokens for communicating with the API
+		prc.jwtTokens     = getInstance( "JWTService@cbsecurity" ).fromUser( prc.oCurrentAuthor );
```
I think the API should be mandatory/accessible for the admin to work, personally.  Long-term, I think we should make the Admin API-driven, and reduce all of the duplication of concern and routines that exist between both the API and Admin  modules.  

We can certainly allow it to run headless and add a setting for an API base URL - which can be used by the JS, but

jclausen commented on modules/contentbox/models/content/RelocationService.cfc:
```diff
@@ -0,0 +1,127 @@
+				}
+			);
+		} else {
+			relocation.setTarget( arguments.target );
```
Good catch. Done.

jclausen commented on .cfmigrations.json:
```diff
@@ -0,0 +1,18 @@
+{
```
That is a great question. I believe they would have to define a `default` manually, and @elpete  can correct me if I'm wrong, because this config supercedes the environment variables.   I know when I tried to run it without a manager specified it tells me that no `default` is present.   We will want to add this to the release docs, as well.

jclausen commented on modules/contentbox/migrations/2022_10_13_133204_CONTENTBOX-1296-Unique-Site-Slug-Relocation.cfc:
```diff
@@ -0,0 +1,26 @@
+component {
```
I don't see that being sustainable, long term, as the version marked may not be the version it actually is released under ( e.g. maybe we need to patch master, but we have already created a migration identified with the patch - which means we have to rename the migrations, which will run them again, etc... )

Since the changelog includes the stories addressed and the story ID is in the migtration, I'm not sure I understand the need for the redundancy.  It doesn't seem agile to me to hard-code version numbers in to the migration files.

jclausen replied:
That said, I have renamed the file to expedite this PR

lmajano replied:
That's a good point.  I guess the OCD in me wants them to be controlled. However, I see your point.ALso, I see the point of this directory becoming HUUGE and unmanageable.  I wonder if we can split it into version buckets.  Even if we have to patch master, it will be released under a semantic version.  So that folder could still stand.  I will test with that approach.

jclausen commented:
I have addressed all of the issues except for the unresolved ones above that require your input/direction.

- See my comment on the API
- See my comment/concern on modifying Coldbox with this version of ContentBox in its current beta phase
- Your note on the docs will have to be addressed separately, since they are in a different repo

jclausen commented on box.json:
```diff
@@ -63,11 +63,11 @@
-        "contentbox:install":"install && cd modules/contentbox && install",
-        "contentbox:migrate:create":"migrate create name=${1} migrationsDirectory=modules/contentbox/migrations",
-        "contentbox:migrate":"migrate up migrationsDirectory=modules/contentbox/migrations",
+        "contentbox:install":"install --force && cd modules/contentbox && install --force",
```
Note:  I added the `--force` flag here because, as we go forward we will need to make sure the module and framework updates for subsequent releases are up to date.  I think it makes sense, if they are running this script to perform an up-to-date install

lmajano replied:
agreed

lmajano commented:
This is now merged.  One observation I had that maybe deserves another ticket, is what if an admin wants to holistically see all the relocations.  Maybe not manage them but at least see all of them as in a report.  Maybe a new `System > Relocations Report` that it can basically present allt he relocations present in the system.  Thoughts  @jclausen

jclausen commented:
I think that's a great enhancement.  I will create a story for that.
